components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java

```diff
@@ -1323,6 +1388,38 @@ public static AuthenticationContext getAuthenticationContextFromCache(String con
         return authenticationContext;
     }
 
+    /**
+     * Retrieve the authentication context from cache. If the cached entry is expired, an exception is thrown.
+     *
+     * @param contextId Context id.
+     * @return Authentication context.
+     * @throws FrameworkClientException
+     */
+    public static AuthenticationContext getAuthenticationContextFromCacheWithExpiry(String contextId)
+            throws FrameworkClientException {
+
+        AuthenticationContext authenticationContext = null;
+        AuthenticationContextCacheKey cacheKey = new AuthenticationContextCacheKey(contextId);
+        AuthenticationContextCacheEntry authenticationContextCacheEntry = AuthenticationContextCache.getInstance().
+                getValueFromCache(cacheKey);
+
+        if (authenticationContextCacheEntry != null) {
+            long expiryTimeNanos = authenticationContextCacheEntry.getCreatedTimeNano() +
+                    authenticationContextCacheEntry.getValidityPeriod();
+            if (FrameworkUtils.getCurrentStandardNano() > expiryTimeNanos) {
+                throw new FrameworkClientException(ERROR_EXPIRED_AUTHENTICATION_CONTEXT.getCode(),
```

Comment:
Not sure FrameworkClientException is the right type to be thrown here as this method could even be called by a server component.

Also for an expired session, can't we simply return null?